### PR TITLE
Responsive Design

### PR DESCRIPTION
**Pull request by 112Nisha**
Frontend of the codebase was modified to make the system user-friendly on a mobile phone
Email functionality was also incorporated to send mails when users are added or removed

### CODE REVIEW

File: backend/controllers/workshop.js [hunk 1]
```diff
@@ -73,11 +73,23 @@ const addTemplates = async (req, res) => {
...
   const row = [r];
   await appendIntoSheet(row, SPREADSHEET_ID, SPREADSHEET_USER_TAB_RANGE);
+  try {
+    const path = require('path');
+    const jsonString = await fs.readFile(path.join(__dirname, '..', 'template_mail.json'), 'utf8');
+    const data = JSON.parse(jsonString);
+    sendmail(row[0][0], data.emails.grant_access);
+  } catch (err) {
+    console.error('Error reading or parsing file:', err);
+    throw err;
```
Comment (raj-vlabs) [inline backend/controllers/workshop.js]: If you throw error here the program would terminate. Not being able to send an email is not a critical failure. Please just log the error and return from the function.

File: backend/template_mail.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
+    "emails": {
+        "grant_access": {
```
Comment (raj-vlabs) [inline backend/template_mail.json]: Please rename the templates to access_grant, access_revoke etc so the context comes first and then the action

File: backend/template_mail.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+            "subject": "Welcome to Virtual Labs IIIT Hyderabad - Workshop Tracker Access",
+            "body": "Dear Nodal Coordinator,\nGreetings from Virtual Labs!\nWe are delighted to welcome your college as a Nodal Centre for Virtual Labs IIIT Hyderabad. Thank you for your interest in this initiative.\nYou have been granted access to our Workshop Tracker. Use the Nodal Coordinator email from your Expression of Interest document as your login credentials.\nThe portal allows you to upload workshop-related data for your Nodal Centre and View other Nodal Centres associated with Virtual Labs IIIT Hyderabad. Please refer the user document for reference.\nFor any queries, please contact us at ioc-iiith@vlabs.ac.in.\nWe look forward to a successful collaboration.",
+            "html": ""
```
Comment (raj-vlabs) [inline backend/template_mail.json:6]: What is the order of processing of body and html. Should there be only one of them in the config?
It should be clearly documented

File: backend/utils/sheet.js [hunk 4]
```diff
@@ -266,49 +266,73 @@ const getInstitutes_ = async () => {
...
-                sheetId: sheetId,
-                dimension: "ROWS",
-                startIndex: rowIndex,
-                endIndex: rowIndex + 1,
+const deleteFromSheet = async (rowIndex, table) => {
+  // console.log(table);
+  const result = getUsers_();
+  result.then(value => {
+  const email = value.rows[rowIndex - 1]['User Email'];
+    (async () => {
```
Comment (raj-vlabs) [inline backend/utils/sheet.js]: Why are we using this IIFE here? The main function is already an async function

File: backend/utils/sheet.js [hunk 5]
```diff
@@ -266,49 +266,73 @@ const getInstitutes_ = async () => {
...
+      try {
+        const path = require('path');
+        const jsonString = await fs.readFile(path.join(__dirname, '..', 'template_mail.json'), 'utf8');
+        const data = JSON.parse(jsonString);
+        sendmail(email, data.emails.revoke_access);
+      } catch (err) {
+        console.error('Error reading or parsing file:', err);
+        throw err;
```
Comment (raj-vlabs) [inline backend/utils/sheet.js]: Log the error and return from here instead of rethrowing it as in the access_grant case

File: docs/tech-doc.md [hunk 6]
```diff
@@ -29,14 +29,18 @@ The Outreach tool utilizes Google OAuth authentication for user authentication.
...
+
 ## Development Environment Setup
 
 Developers can set up their local development environment using the following steps:
 
 1. Clone the repository from GitHub.
 2. Navigate to the project directory.
 3. Navigate to frontend folder. Install frontend dependencies using `npm install` or `yarn install`.
-4. Copy `example.env` to `.env` and update the environment variables.
+4. Copy `example.env` to `.env` and update the environment variables. Ensure the the email account being used here has enabled access to 3rd party applications.
```
Comment (raj-vlabs) [inline docs/tech-doc.md]: Please list all the requirements and configuration required on the gmail account used to send emails

File: frontend/src/nc/css/App.css [hunk 7]
```diff
@@ -406,7 +406,7 @@ html[data-w-dynpage] [data-w-cloak] {
   cursor: pointer;
   box-shadow: 0 0 0 1px rgba(0, 0, 0, .1), 0 1px 3px rgba(0, 0, 0, .1);
   visibility: visible !important;
```
Comment (raj-vlabs) [inline frontend/src/nc/css/App.css:408]: Please use !important judiciously. This should strictly be the last resort

File: frontend/src/nc/pages/Dashboard.jsx [hunk 8]
```diff
@@ -80,7 +80,11 @@ const Dashboard = () => {
           setViewHelp={setViewHelp}
         />
       </div>
-      <div className="flex flex-1 flex-col overflow-hidden">
+      <div className="flex flex-1 flex-col overflow-hidden" style={
+        {
+          overflow: "auto",
```
Comment (raj-vlabs) [inline frontend/src/nc/pages/Dashboard.jsx]: isn't there a built-in construct that handles this case? Do we need to explicitly add an inline style attribute?

File: frontend/src/nc/pages/components/AddModal.js [hunk 9]
```diff
@@ -210,7 +219,7 @@ const AddModal = ({ setModal, table, columns_, postEndpoint, refreshFunc }) => {
             &times;
           </span>
         </div>
-        <form className="flex flex-col mb-4" onSubmit={handleSubmit}>
+        <form className="flex flex-col mb-4 overflow-y-auto overflow-x-hidden max-h-[500px]" onSubmit={handleSubmit}>
```
Comment (raj-vlabs) [inline frontend/src/nc/pages/components/AddModal.js:222]: can you use this kind of an overflow class in place of in-line styling in Dashboard.jsx?

File: frontend/src/nc/pages/components/Dropdowns.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
+import React, { useState } from 'react';
```
Comment (raj-vlabs) [inline frontend/src/nc/pages/components/Dropdowns.js]: Should the extension of this file be jsx?

File: frontend/src/nc/pages/components/AddModal.js [hunk 11]
```diff
@@ -43,6 +43,11 @@ const AddModal = ({ setModal, table, columns_, postEndpoint, refreshFunc }) => {
       const email = JSON.parse(localStorage.getItem("user"))?.email;
```
Comment (raj-vlabs) [inline frontend/src/nc/pages/components/AddModal.js:43]: should the extension of this file be jsx?

File: frontend/src/nc/pages/components/EditModal.js [hunk 12]
```diff
@@ -181,8 +181,8 @@ const EditModal = ({ editObj, setEditObj, columns_, refreshFunc }) => {
   };
```
Comment (raj-vlabs) [inline frontend/src/nc/pages/components/EditModal.js:181]: should the extension of this file be jsx?

**Review by raj-vlabs — CHANGES_REQUESTED** [review]
Please resolve all comments. General comments:
1. Please check for the file extensions between js and jsx
2. what is his.txt file and why is it checked in?

File: backend/controllers/workshop.js (hunk 1 above)
Comment (112Nisha) [reply backend/controllers/workshop.js]: Done

File: backend/template_mail.json (hunk 2 above)
Comment (112Nisha) [reply backend/template_mail.json]: Done

File: backend/utils/sheet.js (hunk 5 above)
Comment (112Nisha) [reply backend/utils/sheet.js]: Done

File: docs/tech-doc.md (hunk 6 above)
Comment (112Nisha) [reply docs/tech-doc.md]: Done

File: frontend/src/nc/pages/Dashboard.jsx (hunk 8 above)
Comment (112Nisha) [reply frontend/src/nc/pages/Dashboard.jsx]: Done

File: frontend/src/nc/pages/components/AddModal.js (hunk 9 above)
Comment (112Nisha) [reply frontend/src/nc/pages/components/AddModal.js:222]: Done

File: backend/template_mail.json (hunk 3 above)
Comment (112Nisha) [reply backend/template_mail.json:6]: body will be processed only if html is an empty string. I have added this information to the tech-doc.

File: frontend/src/nc/pages/components/AddModal.js (hunk 11 above)
Comment (112Nisha) [reply frontend/src/nc/pages/components/AddModal.js:43]: Done

File: frontend/src/nc/pages/components/EditModal.js (hunk 12 above)
Comment (112Nisha) [reply frontend/src/nc/pages/components/EditModal.js:181]: Done

File: backend/utils/sheet.js (hunk 4 above)
Comment (112Nisha) [reply backend/utils/sheet.js]: Right, but in order for the function to execute, result must be processed, and it no longer remains truly asynchronous. In order to keep the mail sending and modification of the sheets separated and asynchronous with each other, I have wrapped it around an async function.

File: frontend/src/nc/css/App.css (hunk 7 above)
Comment (112Nisha) [reply frontend/src/nc/css/App.css:408]: The !important made use of here was not by me.

File: frontend/src/nc/pages/components/Dropdowns.js (hunk 10 above)
Comment (112Nisha) [reply frontend/src/nc/pages/components/Dropdowns.js]: Done. Also, I observed that there are a lot more files in the repo that already make use of react and are still named .js. Do you want me to change the extension of all of them to .jsx?

File: backend/utils/sheet.js (hunk 4 above)
Comment (raj-vlabs) [reply backend/utils/sheet.js]: So I am guessing you want email sending and sheet modification to go in parallel. For that the right approach is that you collect all promises from these functions and wait till all of them are resolved. Take a look [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all)

**Review by raj-vlabs — CHANGES_REQUESTED** [review]

**Comment by 112Nisha** [comment]
In the code I have pushed now, I have made use of Promises for email sending and sheet modification to go in parallel.

**Review by raj-vlabs — APPROVED** [review]

**Pull request merged**